lib/eco/incparser/error_recovery.py

```diff
@@ -0,0 +1,207 @@
+            # Can't recover if EOS caused the error. But why?
+            return False
+
+        # Get the character offset of the top node on the stack
+        error_offset = self.stack_offset(len(self.stack))
+
+        node = self.stack[-1]
+        while not isinstance(node, EOS):
+            if node.new:
+                # Can't recover a newly created subtree
+                self.stack.pop()
+                node = self.stack[-1]
+                continue
+
+            # Check if current node can be isolated
+            if self.is_valid_iso_subtree(node, error_offset):
+                return True
+
+            # Couldn't isolate current node. Try it's parents of the last
+            # successfully parsed tree (including recent changes)
+            while node.get_attr("parent", self.previous_version):
+                node = node.get_attr("parent", self.previous_version)
+                if node is self.root:
+                    break
+                if node in ancestors_to_ignore:
+                    continue
+                # Check if current node can be isolated
+                if self.is_valid_iso_subtree(node, error_offset):
+                    return True
+                else:
+                    ancestors_to_ignore.add(node)
+
+            # Couldn't find elegible parent. Now try more previously parsed subtrees
+            # on the stack
+            self.stack.pop()
+            node = self.stack[-1]
+        return False
+
+    def is_valid_iso_subtree(self, node, error_offset):
+        if node in self.rejects:
+            # Node has already been rejected as a candidate for recovery
+            return False
+
+        self.rejects.add(node)
+
+        if len(node.children) == 0:
+            return False
+
+        cut = self.get_cut(node)
+        if cut == -1:
+            return False
+
+        left_offset = self.stack_offset(cut) # offset of the stack node at position `cut`
+        if left_offset > error_offset:
+            # The node offset is after the error
+            return False
+
+        last_offset = self.offset(node, self.previous_version) # character offset of the isolation node
+
+        # because I am using get_cut on all nodes, get_cut would fail before
+        # this check. Wagner doesn't use get_cut on stack nodes which is why he
+        # needed this check
+        if left_offset != last_offset:
+            # Offsets don't align
+            return False
+
+        # Get the subtrees character length as in the previous version
+        nlen = node.textlength(version = self.previous_version)
+        if last_offset + nlen < error_offset:
+            # The offset range of the isolation tree doesn't meet or exceed the
+            # location of the error
+            return False
+
+        # Reaching this point means we found a candidate for isolation. Now we
+        # need to check if we can continue parsing if we isolate this subtree
+
+        cut_point = cut
```

Comment:
There seems to be no need for a new variable `cut_point`? `cut` isn't referenced after this point in the function as far as I can tell.